physicalplan/function/function.go

```diff
@@ -410,13 +614,46 @@ func maybeSample(lbls labels.Labels, stepTime time.Time, points []promql.Point,
 	}
 }
 
+type valWithScalarFunc func(points []promql.Point, scalars []promql.Point) (float64, bool)
+
+func calculatedSample(lbls labels.Labels, stepTime time.Time, points, scalarPoints []promql.Point, valFunc valWithScalarFunc) func() promql.Sample {
+	return func() promql.Sample {
+		val, ok := valFunc(points, scalarPoints)
+		if !ok {
+			return InvalidSample
+		}
+
+		return promql.Sample{
+			Metric: lbls,
+			Point: promql.Point{
+				T: stepTime.UnixMilli(),
+				V: val,
+			},
+		}
+	}
+}
+
 func ignoreEmpty(points []promql.Point, f func() promql.Sample) promql.Sample {
 	if len(points) == 0 {
 		return InvalidSample
 	}
 	return f()
 }
 
+func ignoreEmptyScalar(points, scalars []promql.Point, f func() promql.Sample) promql.Sample {
```

Comment:
Why do we need this function? Scalars should always be available and the syntax will be validated by the PromQL parser.